File: trunk/src/dbModel/Table.cpp
#include "Table.h"

#include "Schema.h"
#include "TableColumn.h"
#include "PrimaryKey.h"
#include "UniqueConstraint.h"
#include "TextNotEmptyCheckConstraint.h"
#include "ForeignKey.h"

Table::Table(Schema* s, const QString& name)
  : DbInterfacePartInSchema(s, name),
    _primaryKey(0)
{}

Table::~Table()
{}

TableColumn* Table::createTableColumn(const QString& name, Database::DataTypes t) {
  _columns.append(new TableColumn(this, name, t));
  return _columns.last();
}

QList<TableColumn*>::iterator Table::getFirstTableColumn() {
  return _columns.begin();
}

QList<TableColumn*>::iterator Table::getLastTableColumn() {
  return _columns.end();
}

PrimaryKey* Table::createPrimaryKey(const QString& n) {
  _primaryKey = new PrimaryKey(this, n);
  return _primaryKey;
}

UniqueConstraint* Table::createUniqueConstraint(const QString& n) {
  _uniqueConstraints.append(new UniqueConstraint(this, n));
  return _uniqueConstraints.last();
}

TextNotEmptyCheckConstraint* Table::createTextNotEmptyCheckConstraint(const QString& n) {
  _textNotEmptyCheckConstraints.append(new TextNotEmptyCheckConstraint(this, n));
  return _textNotEmptyCheckConstraints.last();
}

QString Table::getQualifiedName() {
   return QString("%1.%2").arg(getSchema()->getName()).arg(getName());
}

ForeignKey* Table::createForeignKey(const QString& name) {
  _foreignKeys.append(new ForeignKey(this, name));
  return _foreignKeys.last();
}

QList<ForeignKey*>::iterator Table::getFirstForeignKey() {
  return _foreignKeys.begin();
}

QList<ForeignKey*>::iterator Table::getLastForeignKey() {
  return _foreignKeys.end();
}

TableColumn* Table::getIdColumn() {
  if (!hasPrimaryKey()) {
    return 0;
  }

  return *(getPrimaryKey()->getFirstColumn());
}

TableColumn* Table::getTableColumn(const QString& name) {
  for (QList<TableColumn*>::iterator it = getFirstTableColumn(); it != getLastTableColumn(); it++) {
    TableColumn* c = *it;
    if (c->getName() == name) {
      return c;
    }
  }
  return 0;
}

File: tools/CodeGenerators/codegen/src/Table.py
from Entity import *
from TableColumn import *
from PrimaryKey import *
from UniqueConstraint import *
from NotEmptyConstraint import *
from ForeignKey import *
from RangeCheckConstraint import *

class Table(Entity):
    def __init__(self, schema, name, 
                 hasIdColumn=True, 
                 hasNameColumn=False, 
                 nameColumnIsUnique=False, 
                 hasDescriptionColumn=False):
        Entity.__init__(self, name)
        self.columns = {}
        self.schema = schema
        self.uniqueConstraints = {}
        self.notEmptyConstraints = {}
        self.rangeCheckConstraints = {}
        self.foreignKeys = {}
        self.primaryKey=None
        if hasIdColumn:
            self.createIdColumn('seq_%s' % self.name)
        if hasNameColumn:
            self.createNameColumn()
        if nameColumnIsUnique:
            self.createUniqueConstraintOnNameColumn()
        if hasDescriptionColumn:
            self.createDescriptionColumn()
    def createColumn(self, name, dataType=None, nullable=False, 
                     sequence=None, defaultValue=None, 
                     defaultText=None, primaryKey=False, 
                     referencedColumn=None, notEmpty=False, isUnique=False):
        self.columns[name] = TableColumn(self, name, dataType, nullable, 
                                         sequence, defaultValue, defaultText, 
                                         primaryKey, referencedColumn, 
                                         notEmpty, isUnique)
        return self.column(name)
    def column(self, name):
        return self.columns[name]
    def createPrimaryKey(self, name, column):
        self.primaryKey = PrimaryKey(self, name, column)
    def hasPrimaryKey(self):
        return self.primaryKey is not None
    def createUniqueConstraint(self, name, columns):
        self.uniqueConstraints[name] = UniqueConstraint(self, name, columns)
        return self.uniqueConstraint(name)
    def createRangeCheckConstraint(self, name, column, min, max):
        self.rangeCheckConstraints[name] = RangeCheckConstraint(self,
                                                                column,
                                                                name,
                                                                min,
                                                                max)
    def uniqueConstraint(self, name):
        return self.uniqueConstraints[name]
    def createNotEmptyConstraint(self, name, columns):
        self.notEmptyConstraints[name] = NotEmptyConstraint(self, name, columns)
        return self.notEmptyConstraint(name)
    def notEmptyConstraint(self, name):
        return self.notEmptyConstraints[name]
    def createForeignKey(self, name, localColumn, referencedColumn):
        self.foreignKeys[name] = ForeignKey(self, name, localColumn, referencedColumn)
        return self.foreignKey(name)
    def foreignKey(self, name):
        return self.foreignKeys[name]
    def createIdColumn(self, sequenceName):
        self.createColumn('id', self.schema.database.model.dataType('Integer'), sequence=self.schema.createSequence(sequenceName), primaryKey=True, nullable=False)
    def createNameColumn(self):
        self.createColumn('name', self.schema.database.model.dataType('Unicode'), 
                          nullable=False, 
                          notEmpty=True, 
                          defaultText='new item')
    def createUniqueConstraintOnNameColumn(self):
        self.createUniqueConstraint('u_%s_name' % self.name, [self.column('name'),])
    def createDescriptionColumn(self):
        self.createColumn('description', self.schema.database.model.dataType('Unicode'), nullable=False, defaultText='')
